File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        // // yah vector wala 

        //  int n1 = s1.length();
        // int n2 = s2.length();
        // if (n1 > n2) return false;
        
        // // Primitive functions jaisa behavior, fixed 26 blocks
        // vector<int> s1_count(26, 0);
        // vector<int> s2_count(26, 0);
        
        // for (int i = 0; i < n1; i++) {
        //     s1_count[s1[i] - 'a']++;
        //     s2_count[s2[i] - 'a']++;
        // }
        
        // if (s1_count == s2_count) return true;
        
        // // Fixed-size window sliding
        // for (int i = n1; i < n2; i++) {
        //     s2_count[s2[i] - 'a']++;      // New character basket mein aaya
        //     s2_count[s2[i - n1] - 'a']--; // Old character basket se gaya
            
        //     if (s1_count == s2_count) return true;
        // }
        // return false;

        // yah wala hash map ka hai

         int n1 = s1.length(), n2 = s2.length();
        if (n1 > n2) return false;

        unordered_map<char, int> s1_map;
        unordered_map<char, int> s2_map;

        for (int i = 0; i < n1; i++) {
            s1_map[s1[i]]++;
            s2_map[s2[i]]++;
        }

        if (s1_map == s2_map) return true;

        for (int i = n1; i < n2; i++) {
            s2_map[s2[i]]++; // Object method internally invoked via operator

            char old_char = s2[i - n1];
            s2_map[old_char]--;
            
            // Map Cleanup to keep comparisons precise
            if (s2_map[old_char] == 0) {
                s2_map.erase(old_char); // Class Method Calling
            }

            if (s1_map == s2_map) return true;
        }
        return false;
    }
};

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        // yah hash map se likh samjana kha kiye

        int n1 = s1.length();
        int n2 = s2.length();
        if (n1 > n2) return false;

        // Class and Objects definition
        HashMap<Character, Integer> s1Map = new HashMap<>();
        HashMap<Character, Integer> s2Map = new HashMap<>();

        for (int i = 0; i < n1; i++) {
            s1Map.put(s1.charAt(i), s1Map.getOrDefault(s1.charAt(i), 0) + 1);
            s2Map.put(s2.charAt(i), s2Map.getOrDefault(s2.charAt(i), 0) + 1);
        }

        if (s1Map.equals(s2Map)) return true;

        for (int i = n1; i < n2; i++) {
            char newChar = s2.charAt(i);
            s2Map.put(newChar, s2Map.getOrDefault(newChar, 0) + 1);

            char oldChar = s2.charAt(i - n1);
            s2Map.put(oldChar, s2Map.get(oldChar) - 1);
            
            // Strict Object clean up rules
            if (s2Map.get(oldChar) == 0) {
                s2Map.remove(oldChar);
            }

            if (s1Map.equals(s2Map)) return true;
        }
        return false;

        // yh wala vector se hua hai fast run time ke liye

        // int n1 = s1.length();
        // int n2 = s2.length();
        // if (n1 > n2) return false;
        
        // // Direct array elements manipulation
        // int[] s1Count = new int[26];
        // int[] s2Count = new int[26];
        
        // for (int i = 0; i < n1; i++) {
        //     s1Count[s1.charAt(i) - 'a']++;
        //     s2Count[s2.charAt(i) - 'a']++;
        // }
        
        // if (Arrays.equals(s1Count, s2Count)) return true;
        
        // for (int i = n1; i < n2; i++) {
        //     s2Count[s2.charAt(i) - 'a']++;
        //     s2Count[s2.charAt(i - n1) - 'a']--;
            
        //     if (Arrays.equals(s1Count, s2Count)) return true;
        // }
        // return false;
    }
};